src/neuralnet/superblock.cpp

```diff
@@ -109,74 +311,228 @@ std::string PackBinarySuperblock(std::string sBlock)
-
-                    if (mag > 0) {
-                        m_cpids.emplace(NN::Cpid::Parse(entry.first.objectID), mag);
-                    } else {
-                        m_network.m_zero_mag_cpid_count++;
-                    }
-                }
+                superblock.m_cpids.Add(
+                    Cpid::Parse(object_id),
+                    std::nearbyint(entry.second.statsvalue.dMag));
+
                 break;
 
             case statsobjecttype::byProject:
-                m_projects.emplace(entry.first.objectID, ProjectStats(
-                    std::round(entry.second.statsvalue.dAvgRAC),
-                    std::round(entry.second.statsvalue.dRAC)));
+                superblock.m_projects.Add(object_id, ProjectStats(
+                    std::nearbyint(entry.second.statsvalue.dAvgRAC),
+                    std::nearbyint(entry.second.statsvalue.dRAC)));
 
                 break;
 
-            case statsobjecttype::byCPIDbyProject:
-                {
-                    std::vector<std::string> vObjectID = split(entry.first.objectID, ",");
-                    Cpid cpid = NN::Cpid::Parse(vObjectID[1]);
-
-                    m_projects[vObjectID[0]].m_cpids.emplace(
-                        std::distance(m_cpids.begin(), m_cpids.find(cpid)),
-                        CpidStats(
-                            std::round(entry.second.statsvalue.dTC),
-                            std::round(entry.second.statsvalue.dRAC),
-                            std::round(entry.second.statsvalue.dMag)));
-                }
-                break;
+            default:
+                // The scraper statistics map orders the entries by "objecttype"
+                // starting with "byCPID" and "byProject". After importing these
+                // sections into the superblock, we can exit this loop.
+                //
+                return superblock;
+        }
+    }
 
-            case statsobjecttype::NetworkWide:
-                m_network.m_average_rac = entry.second.statsvalue.dAvgRAC;
-                m_network.m_rac = entry.second.statsvalue.dRAC;
-                m_network.m_magnitude = std::round(entry.second.statsvalue.dMag);
+    return superblock;
+}
 
-                break;
-        }
+Superblock Superblock::UnpackLegacy(const std::string& packed)
+{
+    // Legacy-packed superblocks always initialize to version 1:
+    Superblock superblock(1);
+    LegacySuperblockParser legacy(packed);
+
+    superblock.m_cpids = legacy.ExtractMagnitudes();
+    superblock.m_projects = legacy.ExtractProjects();
+
+    return superblock;
+}
+
+std::string Superblock::PackLegacy() const
+{
+    std::stringstream out;
+
+    out << "<ZERO>" << m_cpids.Zeros() << "</ZERO>"
+        << "<BINARY>";
+
+    for (const auto& cpid_pair : m_cpids) {
+        uint16_t mag = htobe16(
+            // Allowed magnitude range is 0 to 65535:
+            std::max<uint16_t>(0, std::min<uint16_t>(cpid_pair.second, 65535)));
```

Comment:
This is redundant I think...the magnitude is already stored in a `uint16_t`.

Reply:
Looks that way...

void Superblock::CpidIndex::Add(const Cpid cpid, const uint16_t magnitude)